client/cli/src/runner.rs

```diff
@@ -195,24 +195,24 @@ impl<C: SubstrateCli> Runner<C> {
 		match subcommand {
 			Subcommand::BuildSpec(cmd) => cmd.run(chain_spec, network_config),
 			Subcommand::ExportBlocks(cmd) => {
-				let (client, _, _) = builder(self.config)?;
+				let (client, _, _, _) = builder(self.config)?;
 				run_until_exit(self.tokio_runtime, cmd.run(client, db_config))
 			}
 			Subcommand::ImportBlocks(cmd) => {
-				let (client, _, import_queue) = builder(self.config)?;
+				let (client, _, import_queue, _task_manager) = builder(self.config)?;
```

Comment:
@cecton Would it be better to pass `task_manager` into `run_until_exit` here?

Reply:
I think you are right, it's probably better.

At the moment, the tokio runtime is moved to the arguments of `run_until_exit`, therefore it's dropped right after the call. But we don't call `terminate` at all.

Normally there are no tasks running so I guess that's okay but since we have a task_manager here, we should be defensive and assume that the code could have a background task, so we should drop everything in the right order.

For example you can also pass the task_manager by value to the `run_until_exit` and change the `run_until_exit` to drop the tokio runtime before the end of the function (just like in `run_service_until_exit`).

I'm starting to think that dropping the task manager before calling terminate is an error and should probably panic or at least log an error. :thinking: that would enforce a proper clean-up of the tasks. What do you think?

Reply:
A problem I have with the code at the moment is that the tasks can be terminated either explicitly via `task_manager.terminate()` or implicitly via dropping the task manager. Ideally only one of these to be valid.

I'm _slightly_ leaning towards doing this on `drop` here actually, because while it is not explicit enough, it is enforced by the compiler.

Reply:
:sweat_smile: that's exactly what I am talking about when I say:

> I'm starting to think that dropping the task manager before calling terminate is an error and should probably panic or at least log an error. :thinking: that would enforce a proper clean-up of the tasks. What do you think?

But you say:

> I'm slightly leaning towards doing this on drop here actually, because while it is not explicit enough, it is enforced by the compiler.

This will actually not work. The tasks need to finish in the async runtime while the keep-alive stuff are kept alive. So the killing the task in the drop will never work.

I think the drop should actually panic if terminate hasn't been called.

Reply:
>The tasks need to finish in the async runtime while the keep-alive stuff are kept alive.

Oooh, okay! That makes sense! Yes, let's enforce that `terminate()` is called then.